benchmarks/htap/lib/monitoring.py

```diff
@@ -27,137 +26,124 @@ class Stats:
 
-    def get_latest_delivery_date(self):
-        return self.latest_delivery_date
+        self.data['tpcc_previous'] = deepcopy(self.data['tpcc'])
 
-    def update(self, reporting_queue, time_now):
         while not reporting_queue.empty():
             src, item = reporting_queue.get_nowait()
             self._update_stats(src, item, time_now)
 
-    def consume_stats(self):
-        tpcc_ok_sum, tpcc_err_sum, _ = self.tpcc_totals()
-        tpcc_ok_previous = self.data['tpcc_ok_previous']
-        tpcc_err_previous = self.data['tpcc_err_previous']
-        self.data['tpcc_ok_previous'] = tpcc_ok_sum
-        self.data['tpcc_err_previous'] = tpcc_err_sum
-
-        return (tpcc_ok_sum, tpcc_ok_sum - tpcc_ok_previous,
-                tpcc_err_sum, tpcc_err_sum - tpcc_err_previous)
+        # add tpc history stats
+        self.data['tpcc_tps_history'].append(list(self.tpcc_tps()))
+        self.updates += 1
+        if self.updates % self.tpcc_csv_interval == 0:
+            self.write_tpcc_stats()
 
     def tpcc_stats_for_stream_id(self, stream_id):
         return self.data['tpcc'][stream_id]
 
     def tpch_stats_for_stream_id(self, stream_id):
         return self.data['tpch'][stream_id]
 
-    def tpcc_totals(self):
-        ok_total = 0
-        err_total = 0
-        new_order_total = 0
+    def tpcc_tps_for_slot(self, slot, slot_type):
+        if not isinstance(self.data['tpcc_previous'][slot]['timestamp'], datetime):
+            return 0
+        delta = self.data['tpcc'][slot][slot_type] - self.data['tpcc_previous'][slot][slot_type]
+        dt = (self.data['tpcc'][slot]['timestamp'] - self.data['tpcc_previous'][slot]['timestamp']).total_seconds()
+        return 0 if dt <= 0 else int(delta / dt)
 
-        for tpcc in self.data['tpcc']:
-            ok_total += tpcc['ok_count']
-            err_total += tpcc['err_count']
-            new_order_total += tpcc['new_order_count']
+    def tpcc_tps(self):
+        return (sum(self.tpcc_tps_for_slot(slot, slot_type) for slot in range(self.num_oltp_slots))
+                for slot_type in ['ok_count', 'timeout_count', 'err_count'])
 
-        return ok_total, err_total, new_order_total
+    def tpcc_totals(self):
+        return (sum(slot[slot_type] for slot in self.data['tpcc'])
+                for slot_type in ['ok_count', 'timeout_count', 'err_count', 'new_order_count'])
 
-    def tpch_totals(self):
-        tpch_data = self.data['tpch']
-        return (
-                sum(slot['finished_count'] for slot in tpch_data),
-                sum(slot['error_count'] for slot in tpch_data),
-                sum(slot['timeout_count'] for slot in tpch_data)
-        )
 
+    def tpch_totals(self):
+        return (sum(slot[slot_type] for slot in self.data['tpch'])
+                for slot_type in ['ok_count', 'error_count', 'timeout_count'])
+
+    def write_tpcc_stats(self):
+        last_60 = self.data['tpcc_tps_history'][-60:]
```

Comment:
So even if we want to report tpcc every 10 mins we report only the last minute in csv?

Reply:
yes